src/markdown/generator.ml

```diff
@@ -0,0 +1,327 @@
+      | _ -> Stop_and_keep)
+  in
+  (c, rest)
+
+let heading { Heading.label; level; title } =
+  let level =
+    match level with
+    | 1 -> "#"
+    | 2 -> "##"
+    | 3 -> "###"
+    | 4 -> "####"
+    | 5 -> "#####"
+    | 6 -> "######"
+    | _ -> ""
+    (* We can be sure that h6 will never be exceded! *)
+  in
+  match label with
+  | Some _ -> command level (str " " ++ inline title)
+  | None -> command (level ^ " ") (inline title)
+
+let inline_subpage = function
+  | `Inline | `Open | `Default -> true
+  | `Closed -> false
+
+let item_prop nbsp = String ("###### " ^ nbsp)
+
+let rec documented_src (l : DocumentedSrc.t) nbsp =
+  match l with
+  | [] -> noop
+  | line :: rest -> (
+      let continue r = documented_src r nbsp in
+      match line with
+      | Code _ ->
+          let c, rest = take_code l in
+          source_code c ++ continue rest
+      | Alternative alt -> (
+          match alt with
+          | Expansion { expansion; url; _ } ->
+              if expansion_not_inlined url then
+                let c, rest = take_code l in
+                source_code c ++ continue rest
+              else documented_src expansion nbsp)
+      | Subpage p -> subpage p.content nbsp ++ continue rest
+      | Documented _ | Nested _ ->
+          let lines, _, rest =
+            Take.until l ~classify:(function
+              | DocumentedSrc.Documented { code; doc; anchor; _ } ->
+                  Accum [ (`D code, doc, anchor) ]
+              | DocumentedSrc.Nested { code; doc; anchor; _ } ->
+                  Accum [ (`N code, doc, anchor) ]
+              | _ -> Stop_and_keep)
+          in
+          let f (content, doc, (anchor : Odoc_document.Url.t option)) =
+            let doc = match doc with [] -> noop | doc -> block doc in
+            let content =
+              match content with
+              | `D code -> inline code
+              | `N l -> documented_src l nbsp
+            in
+            let anchor = match anchor with Some a -> a.anchor | None -> "" in
+            break ++ break ++ Anchor anchor ++ break ++ item_prop nbsp
+            ++ content ++ break ++ break ++ str "  " ++ doc ++ break ++ break
+          in
+          let l = list ~sep:noop (List.map f lines) in
+          l ++ continue rest)
+
+and subpage { title = _; header = _; items; url = _ } nbsp =
+  let content = items in
+  let surround body =
+    if content = [] then break else break ++ break ++ body ++ break
+  in
+  surround @@ item nbsp content
+
+and item nbsp (l : Item.t list) : Markup.t =
+  match l with
+  | [] -> noop
+  | i :: rest -> (
+      let continue r = if r = [] then noop else break ++ break ++ item nbsp r in
+      match i with
+      | Text b -> block b ++ continue rest
+      | Heading h -> break ++ heading h ++ break ++ continue rest
+      | Declaration { attr = _; anchor; content; doc } ->
+          let nbsp' = "&nbsp; &nbsp; &nbsp;" in
+          let decl = documented_src content (nbsp ^ nbsp') in
```

Comment:
Instead of a string, `nbsp` could be a `Markup.t` or a `Markup.t list`. (if it makes the code clearer)

Reply:
same as https://github.com/ocaml/odoc/pull/723#discussion_r712841985